pkgs/by-name/li/libfprint-focaltech-2808-a658/package.nix

```diff
@@ -0,0 +1,92 @@
+{
+  stdenv,
+  lib,
+  fetchurl,
+  rpm,
+  cpio,
+  glib,
+  gusb,
+  pixman,
+  libgudev,
+  nss,
+  libfprint,
+  cairo,
+  pkg-config,
+  autoPatchelfHook,
+}:
+
+# https://discourse.nixos.org/t/request-for-libfprint-port-for-2808-a658/55474
+stdenv.mkDerivation rec {
+  pname = "libfprint-focaltech-2808-a658";
+  libso = "libfprint-2.so.2.0.0";
+  version = "1.94.4";
+
+  # https://gitlab.freedesktop.org/libfprint/libfprint/-/merge_requests/413#note_2476573
+  src = fetchurl {
+    url = "https://github.com/ftfpteams/RTS5811-FT9366-fingerprint-linux-driver-with-VID-2808-and-PID-a658/raw/b040ccd953c27e26c1285c456b4264e70b36bc3f/libfprint-2-2-${version}+tod1-FT9366_20240627.x86_64.rpm";
+    hash = "sha256-MRWHwBievAfTfQqjs1WGKBnht9cIDj9aYiT3YJ0/CUM=";
+  };
+
+  nativeBuildInputs = [
+    rpm
+    cpio
+    pkg-config
+    autoPatchelfHook
+  ];
+
+  buildInputs = [
+    stdenv.cc.cc
+    glib
+    gusb
+    pixman
+    nss
+    libgudev
+    libfprint
+    cairo
+  ];
+
+  unpackPhase = ''
+    rpm2cpio $src | cpio -idmv
+  '';
+
+  pout = placeholder "out";
+  pcText = ''
+    prefix=${pout}
+    includedir=${pout}/include
+    libdir=${pout}/lib
+
+    Name: libfprint-2
+    Description: Generic C API for fingerprint reader access
+    Version: ${version}
+    Requires: gio-unix-2.0 >= 2.56, gobject-2.0 >= 2.56
+    Libs: -L${pout}/lib -lfprint-2
+    Cflags: -I${pout}/include/libfprint-2
+  '';
+
+  installPhase = ''
+    install -Dm444 usr/lib64/${libso} -t $out/lib
+
+    # create this symlink as it was there in libfprint
+    ln -s -T $out/lib/${libso} $out/lib/libfprint-2.so
+    ln -s -T $out/lib/${libso} $out/lib/libfprint-2.so.2
+
+    # generate pkg-config file
+    mkdir $out/lib/pkgconfig
+    cat <<EOF >$out/lib/pkgconfig/libfprint-2.pc
+    ${pcText}
+    EOF
```

Comment:
This is better solved using `writeText` and then simply symlinking the generated file where you want it to be.

Reply:
I was the one that added this,
does writeText work with `$out`? we need the final $out for this derivation, which is `placeholder "out"`.
when using writeText the `placeholder "out"` the value of ${pout} will be the out of writeText derivation.

Tried the following

<details>
<summary>Details</summary>

```diff
diff --git a/pkgs/by-name/li/libfprint-focaltech-2808-a658/package.nix b/pkgs/by-name/li/libfprint-focaltech-2808-a658/package.nix
index 7edb09662dd5..cbc30c55f9fe 100644
--- a/pkgs/by-name/li/libfprint-focaltech-2808-a658/package.nix
+++ b/pkgs/by-name/li/libfprint-focaltech-2808-a658/package.nix
@@ -1,6 +1,7 @@
 {
   stdenv,
   lib,
+  pkgs,
   fetchurl,
   rpm,
   cpio,
@@ -50,7 +51,7 @@ stdenv.mkDerivation rec {
   '';
 
   pout = placeholder "out";
-  pcText = ''
+  pcText = pkgs.writeText "libfprint-2.pc" ''
     prefix=${pout}
     includedir=${pout}/include
     libdir=${pout}/lib
@@ -71,9 +72,7 @@ stdenv.mkDerivation rec {
     ln -s -T $out/lib/${libso} $out/lib/libfprint-2.so.2
 
     mkdir $out/lib/pkgconfig
-    cat <<EOF >$out/lib/pkgconfig/libfprint-2.pc
-    ${pcText}
-    EOF
+    ln -s ${pcText} $out/lib/pkgconfig/libfprint-2.pc
 
     # get files from libfprint required to build the package
     cp -r ${libfprint}/lib/girepository-1.0 $out/lib
```

but this results in

```
prefix=/nix/store/74nys1q550zdk5lbms2xfgf0z3n1mc9d-libfprint-2.pc
includedir=/nix/store/74nys1q550zdk5lbms2xfgf0z3n1mc9d-libfprint-2.pc/include
libdir=/nix/store/74nys1q550zdk5lbms2xfgf0z3n1mc9d-libfprint-2.pc/lib

Name: libfprint-2
Description: Generic C API for fingerprint reader access
Version: 1.94.4
Requires: gio-unix-2.0 >= 2.56, gobject-2.0 >= 2.56
Libs: -L/nix/store/74nys1q550zdk5lbms2xfgf0z3n1mc9d-libfprint-2.pc/lib -lfprint-2
Cflags: -I/nix/store/74nys1q550zdk5lbms2xfgf0z3n1mc9d-libfprint-2.pc/include/libfprint-2
```

the correct one should be

```
prefix=/nix/store/l759c3vhkvn34r00bi25r71s3dw3y1y5-libfprint-focaltech-2808-a658-1.94.4
includedir=/nix/store/l759c3vhkvn34r00bi25r71s3dw3y1y5-libfprint-focaltech-2808-a658-1.94.4/include
libdir=/nix/store/l759c3vhkvn34r00bi25r71s3dw3y1y5-libfprint-focaltech-2808-a658-1.94.4/lib

Name: libfprint-2
Description: Generic C API for fingerprint reader access
Version: 1.94.4
Requires: gio-unix-2.0 >= 2.56, gobject-2.0 >= 2.56
Libs: -L/nix/store/l759c3vhkvn34r00bi25r71s3dw3y1y5-libfprint-focaltech-2808-a658-1.94.4/lib -lfprint-2
Cflags: -I/nix/store/l759c3vhkvn34r00bi25r71s3dw3y1y5-libfprint-focaltech-2808-a658-1.94.4/include/libfprint-2
```

that's why I did the cat <<EOF thing
</details>

Reply:
Hm, that's annoying. I think I'd make it templated and then use substituteAll inside the drv. You could even put the pkg-config file into a separate file in that case.

Reply:
Ok there is a `makePkgconfigItem` function

Reply:
Resolved this using `makePkgconfigItem` and `copyPkgconfigItems`. Thanks to @phanirithvij for pointing this out.

Reply:
This is why `placeholder "out"` is working, because all `placeholder "out"` are turned to `@out@` as the first step. source [`pkgs/build-support/make-pkgconfigitem/default.nix`](https://github.com/NixOS/nixpkgs/blob/62494509c9d31a2984c287afba5082c6bae7fcf7/pkgs/build-support/make-pkgconfigitem/default.nix#L21)

There is `substituteAllInPlace` call in `copyPkgconfigItems`, and `substituteAllInPlace` (see [doc](https://nixos.org/manual/nixpkgs/unstable/#fun-substituteAllInPlace)) replaces all `@var@` to `$var` when called. (which we call inside our dervation as a hook)
source: [`pkgs/build-support/setup-hooks/copy-pkgconfig-items.sh`](https://github.com/NixOS/nixpkgs/blob/62494509c9d31a2984c287afba5082c6bae7fcf7/pkgs/build-support/setup-hooks/copy-pkgconfig-items.sh#L42)

So exactly what Atemu said [here](https://github.com/NixOS/nixpkgs/pull/354311#discussion_r1834654158)